File: src/org/jetbrains/plugins/scala/testingSupport/test/TestRunConfigurationForm.java
package org.jetbrains.plugins.scala.testingSupport.test;

import com.intellij.execution.ExecutionBundle;
import com.intellij.execution.configuration.BrowseModuleValueActionListener;
import com.intellij.execution.ui.ClassBrowser;
import com.intellij.execution.ui.ConfigurationModuleSelector;
import com.intellij.ide.util.ClassFilter;
import com.intellij.ide.util.PackageChooserDialog;
import com.intellij.openapi.fileChooser.FileChooserDescriptor;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiPackage;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.ui.RawCommandLineEditor;
import org.jetbrains.plugins.scala.lang.psi.ScalaPsiUtil;
import org.jetbrains.plugins.scala.lang.psi.impl.ScalaPsiManager;

import javax.swing.*;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;

/**
 * @author Ksenia.Sautina
 * @since 5/15/12
 */
public class TestRunConfigurationForm{
  private JPanel myPanel;
  private TextFieldWithBrowseButton testClassTextField;
  private RawCommandLineEditor VMParamsTextField;
  private RawCommandLineEditor testOptionsTextField;
  private TextFieldWithBrowseButton testPackageTextField;
  private JLabel testClassLabel;
  private JLabel testPackageLabel;
  private JComboBox moduleComboBox;
  private TextFieldWithBrowseButton workingDirectoryField;
  private JPanel searchForTestsPanel;
  private ConfigurationModuleSelector myModuleSelector;
  private String suitePath;

  private JComboBox searchForTestsComboBox;
  public static enum SearchForTest {
    IN_WHOLE_PROJECT, IN_SINGLE_MODULE, ACCROSS_MODULE_DEPENDENCIES;

    @Override
    public String toString() {
      switch (this) {
        case ACCROSS_MODULE_DEPENDENCIES: return "Across module dependencies";
        case IN_WHOLE_PROJECT: return "In whole project";
        case IN_SINGLE_MODULE: return "In single module";
        default: return "";
      }
    }
  }

  private JComboBox kindComboBox;
  private JTextField testNameTextField;
  private JLabel testNameLabel;
  private JCheckBox myShowProgressMessagesCheckBox;

  public static enum TestKind {
    ALL_IN_PACKAGE, CLASS, TEST_NAME;

    @Override
    public String toString() {
      switch (this) {
        case ALL_IN_PACKAGE: return "All in package";
        case CLASS: return "Class";
        case TEST_NAME: return "Test name";
        default: return "";
      }
    }

    public static TestKind fromString(String s) {
      if (s.equals("All in package")) return ALL_IN_PACKAGE;
      else if (s.equals("Class")) return CLASS;
      else if (s.equals("Test name")) return TEST_NAME;
      else return null;
    }
  }


  public TestRunConfigurationForm(final Project project, final AbstractTestRunConfiguration configuration) {
    myModuleSelector = new ConfigurationModuleSelector(project, moduleComboBox);
    myModuleSelector.reset(configuration);
    moduleComboBox.setEnabled(true);
    addClassChooser("Choose test class", testClassTextField, project);
    addFileChooser("Choose Working Directory", workingDirectoryField, project);
    VirtualFile baseDir = project.getBaseDir();
    String path = baseDir != null ? baseDir.getPath() : "";
    workingDirectoryField.setText(path);
    addPackageChooser(testPackageTextField, project);
    VMParamsTextField.setDialogCaption("VM parameters editor");
    testOptionsTextField.setDialogCaption("Additional options editor");

    for (SearchForTest searchForTest : SearchForTest.values()) {
      searchForTestsComboBox.addItem(searchForTest);
    }

    searchForTestsComboBox.setSelectedItem(configuration.getSearchTest());

    searchForTestsComboBox.addItemListener(new ItemListener() {
      public void itemStateChanged(ItemEvent e) {
        setupModuleComboBox();
      }
    });

    myShowProgressMessagesCheckBox.setSelected(configuration.getShowProgressMessages());

    for (TestKind testKind : TestKind.values()) {
      kindComboBox.addItem(testKind);
    }

    switch (configuration.getTestKind()) {
      case ALL_IN_PACKAGE:
        setPackageEnabled();
        break;
      case CLASS:
        setClassEnabled();
        break;
      case TEST_NAME:
        setTestNameEnabled();
        break;
    }

    kindComboBox.addItemListener(new ItemListener() {
      public void itemStateChanged(ItemEvent e) {
        moduleComboBox.setEnabled(true);
        switch ((TestKind) e.getItem()) {
          case ALL_IN_PACKAGE:
            setPackageEnabled();
            setupModuleComboBox();
            break;
          case CLASS:
            setClassEnabled();
            break;
          case TEST_NAME:
            setTestNameEnabled();
            break;
        }
      }
    });

    suitePath = configuration.suitePath();
 }

  private void setupModuleComboBox() {
    switch ((SearchForTest) searchForTestsComboBox.getSelectedItem()) {
      case IN_WHOLE_PROJECT:
        moduleComboBox.setEnabled(false);
        break;
      case IN_SINGLE_MODULE:
        moduleComboBox.setEnabled(true);
        break;
      case ACCROSS_MODULE_DEPENDENCIES:
        moduleComboBox.setEnabled(true);
        break;
    }
  }

  private void setPackageVisible(boolean visible) {
    testPackageLabel.setVisible(visible);
    testPackageTextField.setVisible(visible);
    searchForTestsPanel.setVisible(visible);
  }

  private void setClassVisible(boolean visible) {
    testClassLabel.setVisible(visible);
    testClassTextField.setVisible(visible);
  }

  private void setTestNameVisible(boolean visible) {
    testNameLabel.setVisible(visible);
    testNameTextField.setVisible(visible);
    testClassLabel.setVisible(visible);
    testClassTextField.setVisible(visible);
  }

  private void disableAll() {
    setPackageVisible(false);
    setClassVisible(false);
    setTestNameVisible(false);
  }

  private void setPackageEnabled() {
    disableAll();
    setPackageVisible(true);
    kindComboBox.setSelectedItem(TestKind.ALL_IN_PACKAGE);
  }

  private void setClassEnabled() {
    disableAll();
    setClassVisible(true);
    kindComboBox.setSelectedItem(TestKind.CLASS);
  }

  private void setTestNameEnabled() {
    disableAll();
    setTestNameVisible(true);
    kindComboBox.setSelectedItem(TestKind.TEST_NAME);
  }

  public void apply(AbstractTestRunConfiguration configuration) {
    setTestClassPath(configuration.getTestClassPath());
    setJavaOptions(configuration.getJavaOptions());
    setTestArgs(configuration.getTestArgs());
    setTestPackagePath(configuration.getTestPackagePath());
    switch (configuration.getTestKind()) {
      case ALL_IN_PACKAGE:
        setPackageEnabled();
        break;
      case CLASS:
        setClassEnabled();
        break;
      case TEST_NAME:
        setTestNameEnabled();
        break;
    }
    setWorkingDirectory(configuration.getWorkingDirectory());
    myModuleSelector.applyTo(configuration);
    searchForTestsComboBox.setSelectedItem(configuration.getSearchTest());
    setTestName(configuration.getTestName());
    setShowProgressMessages(configuration.getShowProgressMessages());
  }

  public TestKind getSelectedKind() {
    return (TestKind) kindComboBox.getSelectedItem();
  }

  public SearchForTest getSearchForTest() {
    return (SearchForTest) searchForTestsComboBox.getSelectedItem();
  }

  public String getTestClassPath() {
    return testClassTextField.getText();
  }

  public String getTestArgs() {
    return testOptionsTextField.getText();
  }

  public String getJavaOptions() {
    return VMParamsTextField.getText();
  }

  public String getTestPackagePath() {
    return testPackageTextField.getText();
  }

  public String getWorkingDirectory() {
    return workingDirectoryField.getText();
  }

  public void setTestClassPath(String s) {
    testClassTextField.setText(s);
  }

  public void setTestArgs(String s) {
    testOptionsTextField.setText(s);
  }

  public void setJavaOptions(String s) {
    VMParamsTextField.setText(s);
  }

  public void setTestPackagePath(String s) {
    testPackageTextField.setText(s);
  }

  public void setWorkingDirectory(String s) {
    workingDirectoryField.setText(s);
  }

  public String getTestName() {
    return testNameTextField.getText();
  }

  public void setTestName(String s) {
    testNameTextField.setText(s);
  }

  public boolean getShowProgressMessages() {
    return myShowProgressMessagesCheckBox.isSelected();
  }

  public void setShowProgressMessages(boolean b) {
    myShowProgressMessagesCheckBox.setSelected(b);
  }

  public JPanel getPanel() {
    return myPanel;
  }

  private void addClassChooser(final String title,
                               final TextFieldWithBrowseButton textField,
                               final Project project) {
    ClassBrowser browser = new ClassBrowser(project, title) {
      protected ClassFilter.ClassFilterWithScope getFilter() throws ClassBrowser.NoFilterException {
        return new ClassFilter.ClassFilterWithScope() {
          public GlobalSearchScope getScope() {
            Module module = getModule();
            if (module != null) return GlobalSearchScope.moduleWithDependenciesAndLibrariesScope(module);
            return GlobalSearchScope.allScope(project);
          }

          public boolean isAccepted(PsiClass aClass) {
            if (!getScope().accept(aClass.getContainingFile().getVirtualFile())) return false;
            PsiClass[] classes = ScalaPsiManager.instance(project).getCachedClasses(getScope(), suitePath);
            for (PsiClass psiClass : classes) {
              if (ScalaPsiUtil.cachedDeepIsInheritor(aClass, psiClass)) return true;
            }
            return false;
          }
        };
      }

      protected PsiClass findClass(String className) {
        return ScalaPsiManager.instance(project).getCachedClass(GlobalSearchScope.allScope(project), className);
      }
    };

    browser.setField(textField);
  }

  private FileChooserDescriptor addFileChooser(final String title,
                                               final TextFieldWithBrowseButton textField,
                                               final Project project) {
    final FileChooserDescriptor fileChooserDescriptor = new FileChooserDescriptor(false, true, false, false, false, false) {
      @Override
      public boolean isFileVisible(VirtualFile file, boolean showHiddenFiles) {
        return super.isFileVisible(file, showHiddenFiles) && file.isDirectory();
      }
    };
    fileChooserDescriptor.setTitle(title);
    textField.addBrowseFolderListener(title, null, project, fileChooserDescriptor);
    return fileChooserDescriptor;
  }

  private void addPackageChooser(final TextFieldWithBrowseButton textField, final Project project) {
    PackageChooserActionListener browser = new PackageChooserActionListener(project);
    browser.setField(textField);
  }

  //todo: copied from JUnitConfigurable
  private static class PackageChooserActionListener extends BrowseModuleValueActionListener {
    public PackageChooserActionListener(final Project project) {super(project);}

    protected String showDialog() {
      final PackageChooserDialog dialog = new PackageChooserDialog(ExecutionBundle.message("choose.package.dialog.title"), getProject());
      dialog.show();
      final PsiPackage aPackage = dialog.getSelectedPackage();
      return aPackage != null ? aPackage.getQualifiedName() : null;
    }
  }

  public Module getModule() {
    return myModuleSelector.getModule();
  }
}

File: scala/scala-impl/src/org/jetbrains/plugins/scala/testingSupport/test/ui/TestRunConfigurationForm.scala
package org.jetbrains.plugins.scala.testingSupport.test.ui

import java.awt._
import java.util

import com.intellij.application.options.ModuleDescriptionsComboBox
import com.intellij.execution.{ExecutionBundle, ShortenCommandLine}
import com.intellij.execution.configuration.BrowseModuleValueActionListener
import com.intellij.execution.ui.{ClassBrowser, ConfigurationModuleSelector, DefaultJreSelector, JrePathEditor, ShortenCommandLineModeCombo}
import com.intellij.ide.util.{ClassFilter, PackageChooserDialog}
import com.intellij.openapi.editor.ex.EditorEx
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.{ComboBox, ComponentWithBrowseButton, LabeledComponent}
import com.intellij.psi.PsiClass
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.ui.{EditorTextField, EditorTextFieldWithBrowseButton, EnumComboBoxModel, IdeBorderFactory}
import com.intellij.uiDesigner.core.{GridConstraints, Spacer}
import com.intellij.util.ui.{JBUI, UIUtil}
import javax.swing._
import org.jetbrains.annotations.Nls
import org.jetbrains.plugins.scala.ScalaBundle
import org.jetbrains.plugins.scala.extensions.IteratorExt
import org.jetbrains.plugins.scala.lang.psi.ScalaPsiUtil
import org.jetbrains.plugins.scala.lang.psi.impl.ScalaPsiManager
import org.jetbrains.plugins.scala.settings.SimpleMappingListCellRenderer
import org.jetbrains.plugins.scala.testingSupport.test._
import org.jetbrains.plugins.scala.testingSupport.test.testdata._
import org.jetbrains.plugins.scala.testingSupport.test.ui.TestRunConfigurationForm.TextWithMnemonic.Mnemonic
import org.jetbrains.plugins.scala.testingSupport.test.ui.TestRunConfigurationForm.{PackageChooserActionListener, _}
import org.jetbrains.sbt.settings.SbtSettings

import scala.collection.JavaConverters.asScalaBufferConverter

//noinspection ConvertNullInitializerToUnderscore
final class TestRunConfigurationForm(val myProject: Project) {

  private var myWholePanel: JPanel                      = null
  def getPanel: JPanel = myWholePanel

  private var mySuitePaths    : Seq[String]                 = null
  private var myModuleSelector: ConfigurationModuleSelector = null

  // TOP PANEL: TestKind-specific
  private var myTestKind      : LabeledComponent[ComboBox[TestKind]]                                  = null
  private var myClass         : LabeledComponent[EditorTextFieldWithBrowseButton]                     = null
  private var myTestName      : LabeledComponent[TestRunConfigurationForm.MyMultilineEditorTextField] = null
  private var myPackage       : LabeledComponent[EditorTextFieldWithBrowseButton]                     = null
  private var mySearchForTest : LabeledComponent[JComboBox[SearchForTest]]                            = null
  private var myRegex         : LabeledComponent[RegexpPanel]                                         = null
  private var myUseSbtCheckBox: JCheckBox                                                             = null
  private var myUseUiWithSbt  : JCheckBox                                                             = null

  // BOTTOM PANEL: common options
  private var myCommonScalaParameters       : CommonScalaParametersPanel                    = null
  private var myModule                      : LabeledComponent[ModuleDescriptionsComboBox]  = null
  private var myJrePathEditor               : JrePathEditor                                 = null
  private var myShortenClasspathMode        : LabeledComponent[ShortenCommandLineModeCombo] = null
  private var myShowProgressMessagesCheckBox: JCheckBox                                     = null

  init()

  def init(): Unit = {
    createUiComponents()

    myTestKind.component.setModel(new EnumComboBoxModel[TestKind](classOf[TestKind]))
    myTestKind.component.setRenderer(SimpleMappingListCellRenderer.create(
      TestKind.ALL_IN_PACKAGE -> ScalaBundle.message("test.run.config.test.kind.all.in.package"),
      TestKind.CLAZZ -> ScalaBundle.message("test.run.config.test.kind.class"),
      TestKind.TEST_NAME -> ScalaBundle.message("test.run.config.test.kind.test.name"),
      TestKind.REGEXP -> ScalaBundle.message("test.run.config.test.kind.regular.expression")
    ))

    mySearchForTest.component.setModel(new EnumComboBoxModel[SearchForTest](classOf[SearchForTest]))
    mySearchForTest.component.setRenderer(SimpleMappingListCellRenderer.create(
      SearchForTest.IN_WHOLE_PROJECT -> ScalaBundle.message("test.run.config.search.scope.in.whole.project"),
      SearchForTest.IN_SINGLE_MODULE -> ScalaBundle.message("test.run.config.search.scope.in.single.module"),
      SearchForTest.ACCROSS_MODULE_DEPENDENCIES -> ScalaBundle.message("test.run.config.search.scope.across.module.dependencies")
    ))

    myClass.getComponent.setBrowser(new MyClassBrowser(myProject))
    myPackage.getComponent.setBrowser(new PackageChooserActionListener(myProject))

    myJrePathEditor.setDefaultJreSelector(DefaultJreSelector.fromModuleDependencies(myModule.getComponent, false))
    myModuleSelector = new ConfigurationModuleSelector(myProject, myModule.getComponent)

    myTestKind.component.addItemListener(_ => UiComponentsVisibility.update())
    mySearchForTest.component.addItemListener(_ => UiComponentsVisibility.update())
    myUseSbtCheckBox.addItemListener(_ => UiComponentsVisibility.update())

    UiComponentsVisibility.update()
  }

  private def isModuleSelectionSupported(testKind: TestKind, searchScope: SearchForTest): Boolean =
    (testKind, searchScope) match {
      case (TestKind.ALL_IN_PACKAGE, SearchForTest.IN_WHOLE_PROJECT) => false
      case _                                                         => true
    }

  def resetFrom(configuration: AbstractTestRunConfiguration): Unit = {
    val configurationData = configuration.testConfigurationData
    configurationData match {
      case data: AllInPackageTestData => setTestPackagePath(data.testPackagePath)
      case data: RegexpTestData       => setRegexps(data.regexps)
      case data: ClassTestData        =>
        data match {
          case singleTest: SingleTestData =>  setTestName(singleTest.getTestName)
          case _=>
        }
        setTestClassPath(data.testClassPath)
      case _ =>
    }

    setTestKind(configurationData.getKind)
    setSearchForTest(configurationData.getSearchTest)

    resetSbtOptionsFrom(configuration)

    myCommonScalaParameters.reset(configurationData)
    myModuleSelector.reset(configuration)
    myJrePathEditor.setPathOrName(configurationData.getJrePath, true)
    setShortenCommandLine(configuration.getShortenCommandLine)
    setShowProgressMessages(configurationData.getShowProgressMessages)

    mySuitePaths = configuration.javaSuitePaths.asScala

    UiComponentsVisibility.update(configuration)
  }

  def applyTo(configuration: AbstractTestRunConfiguration): Unit = {
    configuration.setTestKind(this.getTestKind)
    configuration.setModule(this.getModule)
    configuration.testConfigurationData = TestConfigurationData.createFromForm(this, configuration)
    configuration.testConfigurationData.initWorkingDirIfEmpty()
    configuration.setShortenCommandLine(getShortenCommandLine)
  }

  private def resetSbtOptionsFrom(configuration: AbstractTestRunConfiguration): Unit = {
    val configurationData = configuration.testConfigurationData
    setUseSbt(configurationData.useSbt)
    setUseUiWithSbt(configurationData.useUiWithSbt)
  }

  def getModule: Module = myModuleSelector.getModule
  def getJrePath: String = myJrePathEditor.getJrePathOrName

  def getTestKind: TestKind = myTestKind.getComponent.getSelectedItem.asInstanceOf[TestKind]
  def getTestClassPath: String = myClass.getComponent.getText
  def getTestName: String = myTestName.getComponent.getText
  def getTestPackagePath: String = myPackage.getComponent.getText
  def getSearchForTest: SearchForTest = mySearchForTest.getComponent.getSelectedItem.asInstanceOf[SearchForTest]
  def getRegexps: (Array[String], Array[String]) = myRegex.getComponent.getRegexps
  def getUseSbt: Boolean = myUseSbtCheckBox.isSelected
  def getUseUiWithSbt: Boolean = myUseUiWithSbt.isSelected
  def getJavaOptions: String = myCommonScalaParameters.getVMParameters
  def getEnvironmentVariables: util.Map[String, String] = myCommonScalaParameters.getEnvironmentVariables
  def getWorkingDirectory: String = myCommonScalaParameters.getWorkingDirectoryAccessor.getText
  def getTestArgs: String = myCommonScalaParameters.getProgramParameters
  def getShowProgressMessages: Boolean = myShowProgressMessagesCheckBox.isSelected
  def getShortenCommandLine: ShortenCommandLine = myShortenClasspathMode.getComponent.getSelectedItem

  private def setTestKind(kind: TestKind): Unit = myTestKind.getComponent.setSelectedItem(kind)
  private def setTestClassPath(s: String): Unit = myClass.getComponent.setText(s)
  private def setTestName(s: String): Unit = myTestName.getComponent.setText(s)
  private def setTestPackagePath(s: String): Unit = myPackage.getComponent.setText(s)
  private def setSearchForTest(searchTest: SearchForTest): Unit = mySearchForTest.getComponent.setSelectedItem(searchTest)
  private def setRegexps(classRegexps: Array[String], testRegexps: Array[String]): Unit = myRegex.getComponent.setRegexps(classRegexps, testRegexps)
  private def setRegexps(regexps: (Array[String], Array[String])): Unit = setRegexps(regexps._1, regexps._2)
  private def setUseSbt(b: Boolean): Unit = myUseSbtCheckBox.setSelected(b)
  private def setUseUiWithSbt(b: Boolean): Unit = myUseUiWithSbt.setSelected(b)
  private def setShowProgressMessages(b: Boolean): Unit = myShowProgressMessagesCheckBox.setSelected(b)
  private def setShortenCommandLine(mode: ShortenCommandLine): Unit = myShortenClasspathMode.getComponent.setSelectedItem(mode)

  private object UiComponentsVisibility {

    def update(): Unit = {
      val testKind = getTestKind
      hideAll()
      testKind match {
        case TestKind.CLAZZ          => setClassVisible(true)
        case TestKind.TEST_NAME      => setTestNameVisible(true)
        case TestKind.REGEXP         => setRegexpVisible(true)
        case TestKind.ALL_IN_PACKAGE => setPackageVisible(true)
      }

      val packageSearchScope = getSearchForTest
      myModule.setEnabled(isModuleSelectionSupported(testKind, packageSearchScope))
      myUseUiWithSbt.setEnabled(myUseSbtCheckBox.isSelected)
    }

    def update(configuration: AbstractTestRunConfiguration): Unit = {
      update()
      val projectHasSbt = hasSbt(myProject)
      myUseSbtCheckBox.setVisible(projectHasSbt) // TODO: hide if sbt support is null/None
      myUseUiWithSbt.setVisible(projectHasSbt && configuration.sbtSupport.allowsSbtUiRun)
    }

    private def hideAll(): Unit = {
      setPackageVisible(false)
      setClassVisible(false)
      setTestNameVisible(false)
      setRegexpVisible(false)
    }

    private def setClassVisible(visible: Boolean): Unit =
      myClass.setVisible(visible)

    private def setTestNameVisible(visible: Boolean): Unit = {
      myTestName.setVisible(visible)
      myClass.setVisible(visible)
    }

    private def setPackageVisible(visible: Boolean): Unit = {
      myPackage.setVisible(visible)
      mySearchForTest.setVisible(visible)
    }

    private def setRegexpVisible(visible: Boolean): Unit =
      myRegex.setVisible(visible)
  }

  private def createUiComponents(): Unit = {
    myWholePanel = new JPanel

    implicit val uiBuilder: GridUiBuilder = new GridUiBuilder(myWholePanel, columns = 4)
    import uiBuilder._

    myTestKind = labeledComponent(ScalaBundle.message("test.run.config.test.kind"), new ComboBox[TestKind])

    myUseSbtCheckBox = (new JCheckBox).setTextWithMnemonic(ScalaBundle.message("test.run.config.use.sbt"))
    myUseUiWithSbt = (new JCheckBox).setTextWithMnemonic(ScalaBundle.message("test.run.config.use.ui.with.sbt"))

    appendRow(myTestKind, myUseSbtCheckBox, myUseUiWithSbt, new Spacer())

    myClass = append(labeledComponent(
      ScalaBundle.message("test.run.config.test.class"),
      new EditorTextFieldWithBrowseButton(myProject, true)
    ))
    myTestName = append(labeledComponent(
      ScalaBundle.message("test.run.config.test.name"),
      new TestRunConfigurationForm.MyMultilineEditorTextField, isMultilineContent = true
    ))
    myPackage = append(labeledComponent(
      ScalaBundle.message("test.run.config.test.package"),
      new EditorTextFieldWithBrowseButton(myProject, false)
    ))
    mySearchForTest = labeledComponent(
      ScalaBundle.message("test.run.config.search.for.tests"),
      new ComboBox[SearchForTest]
    )
    appendRow(mySearchForTest, new Spacer(), new Spacer())
    //    append(mySearchForTest)
    myRegex = append(labeledComponent(
      ScalaBundle.message("test.run.config.regular.expressions"),
      new RegexpPanel, isMultilineContent = true
    ))

    addSeparatorBetweenTopAndBottomPanel()

    myCommonScalaParameters = append(new CommonScalaParametersPanel)
    myModule = append(labeledComponent(
      ExecutionBundle.message("application.configuration.use.classpath.and.jdk.of.module.label"),
      new ModuleDescriptionsComboBox
    ))
    myJrePathEditor = append(new JrePathEditor)
    myShortenClasspathMode = append(labeledComponent(
      ExecutionBundle.message("application.configuration.shorten.command.line.label"),
      new ShortenCommandLineModeCombo(myProject, myJrePathEditor, myModule.getComponent())
    ))
    myShowProgressMessagesCheckBox = (new JCheckBox)
      .setTextWithMnemonic(ScalaBundle.message("test.run.config.print.information.messages.to.console"))
    append(myShowProgressMessagesCheckBox)

    addBottomFiller()

    val topPanelAnchor = myRegex.getLabel
    mySearchForTest.setAnchor(topPanelAnchor)
    myTestKind.setAnchor(topPanelAnchor)
    myClass.setAnchor(topPanelAnchor)
    myTestName.setAnchor(topPanelAnchor)
    myPackage.setAnchor(topPanelAnchor)

    myJrePathEditor.setAnchor(myModule.getLabel)
    myCommonScalaParameters.setAnchor(myModule.getLabel)
    myShortenClasspathMode.setAnchor(myModule.getLabel)
  }

  private def labeledComponent[T <: JComponent](@Nls labelText: String, component: T): LabeledComponent[T] =
    labeledComponent(labelText, component, isMultilineContent = false)

  private def labeledComponent[T <: JComponent](@Nls labelText: String, component: T, isMultilineContent: Boolean): LabeledComponent[T] = {
    val result = new LabeledComponent[T]
    result.setText(labelText)
    result.setComponent(component)
    result.setLabelLocation(BorderLayout.WEST)
    if (isMultilineContent) {
      // by default LabeledComponent's label is aligned in center, so if the inner component is multiline
      // the label has a gab in the TOP and BOTTOM. Moreover, when the inner component grows in height
      // the label "jumps" below it's previous position, which looks awkward
      // This is a workaround to make label position fixed in the TOP
      val TOP_PADDING = 5
      val label = result.getLabel
      label.setVerticalAlignment(SwingConstants.TOP)
      label.setBorder(IdeBorderFactory.createEmptyBorder(JBUI.insetsTop(TOP_PADDING)))
    }
    result
  }

  private def addSeparatorBetweenTopAndBottomPanel()(implicit uiBuilder: GridUiBuilder) = {
    val dim = new Dimension(-1, 32)
    import uiBuilder._
    append(new Spacer, new GridConstraints(
      rowIdx, 0, 1, columns,
      GridConstraints.ANCHOR_CENTER, GridConstraints.FILL_NONE,
      GridConstraints.SIZEPOLICY_FIXED, GridConstraints.SIZEPOLICY_FIXED,
      dim, dim, dim, 0, false
    ))
    append(new JSeparator(SwingConstants.HORIZONTAL))
  }

  private def addBottomFiller()(implicit uiBuilder: GridUiBuilder): Unit = {
    import uiBuilder._
    append(new Spacer, new GridConstraints(
      rowIdx, 0, 1, columns,
      GridConstraints.ANCHOR_CENTER, GridConstraints.FILL_VERTICAL,
      GridConstraints.SIZEPOLICY_WANT_GROW | GridConstraints.SIZEPOLICY_CAN_GROW,
      GridConstraints.SIZEPOLICY_WANT_GROW | GridConstraints.SIZEPOLICY_CAN_GROW,
      null, null, null, 0, false
    ))
  }

  private class MyClassBrowser[T <: JComponent](project: Project)
    extends ClassBrowser[T](project, ScalaBundle.message("test.run.config.choose.test.class")) {

    override protected def getFilter: ClassFilter.ClassFilterWithScope = new ClassFilter.ClassFilterWithScope() {
      override def getScope: GlobalSearchScope =
        getModule match {
          case null   => GlobalSearchScope.allScope(project)
          case module => GlobalSearchScope.moduleWithDependenciesAndLibrariesScope(module)
        }

      override def isAccepted(aClass: PsiClass): Boolean = {
        if (!getScope.accept(aClass.getContainingFile.getVirtualFile)) return false
        val baseSuite = findBaseSuiteClass(project, aClass)
        baseSuite.isDefined
      }

      private def findBaseSuiteClass(project: Project, aClass: PsiClass): Option[PsiClass] =
        (for {
          suitePath  <- mySuitePaths.iterator
          suiteClass <- ScalaPsiManager.instance(project).getCachedClasses(getScope, suitePath).iterator
          if ScalaPsiUtil.isInheritorDeep(aClass, suiteClass)
        } yield suiteClass).headOption
    }

    override protected def findClass(className: String): PsiClass = {
      val cachedClass = ScalaPsiManager.instance(project).getCachedClass(GlobalSearchScope.allScope(project), className)
      if (cachedClass.isEmpty) null
      else cachedClass.get
    }
  }
}

private object TestRunConfigurationForm {

  /** copied from [[com.intellij.execution.junit2.configuration.JUnitConfigurable]] */
  private class PackageChooserActionListener[T <: JComponent](val project: Project)
    extends BrowseModuleValueActionListener[T](project) {

    override protected def showDialog: String = {
      val dialog = new PackageChooserDialog(ExecutionBundle.message("choose.package.dialog.title"), getProject)
      dialog.show()
      val aPackage = dialog.getSelectedPackage
      if (aPackage != null) aPackage.getQualifiedName
      else null
    }
  }

  private class MyMultilineEditorTextField() extends EditorTextField {
    this.setOneLineMode(false)
    override protected def updateBorder(editor: EditorEx): Unit = setupBorder(editor) // in base class border isn't set in multiline mode
  }

  /** based on [[LabeledComponent.TextWithMnemonic]] */
  case class TextWithMnemonic(text: String, mnemonic: Option[Mnemonic]) {
    def setTo(button: AbstractButton): Unit = {
      button.setText(text)
      mnemonic match {
        case Some(Mnemonic(char, index)) =>
          button.setMnemonic(char)
          button.setDisplayedMnemonicIndex(index)
        case _ =>
      }
    }
  }

  object TextWithMnemonic {
    case class Mnemonic(char: Char, index: Int)

    def apply(text: String): TextWithMnemonic = {
      val idx = UIUtil.getDisplayMnemonicIndex(text)
      if (idx != -1)
        new TextWithMnemonic(new StringBuilder(text).deleteCharAt(idx).toString, Some(Mnemonic(text.charAt(idx), idx)))
      else
        new TextWithMnemonic(text, None)
    }
  }

  implicit class AbstractButtonOps[T <: AbstractButton](private val button: T) extends AnyVal {
    def setTextWithMnemonic(text: String): T = {
      TextWithMnemonic(text).setTo(button)
      button
    }
  }
  implicit class EditorTextFieldWithBrowseButtonOps[C <: JComponent](private val target: EditorTextFieldWithBrowseButton) extends AnyVal {
    def setBrowser[B](browser: B)(implicit ev: B <:< BrowseModuleValueActionListener[C]): Unit =
      browser.setField(target.asInstanceOf[ComponentWithBrowseButton[C]])
  }
  implicit class LabeledComponentOps[C <: JComponent](private val target: LabeledComponent[C]) extends AnyVal {
    def component: C = target.getComponent
  }

  private def hasSbt(project: Project): Boolean = {
    val sbtSettings = SbtSettings.getInstance(project)
    val sbtSettingsEmpty = sbtSettings == null || sbtSettings.getLinkedProjectsSettings.isEmpty
    !sbtSettingsEmpty
  }
}
